pages/dkp/kommander/1.3/projects/platform-services/platform-services-catalog/kudo-zookeeper/index.md

```diff
@@ -0,0 +1,184 @@
+```
+INSTANCE=zookeeper;
+kubectl -n test-project-zc6tc get instances -o json | jq ".items[] | select(.metadata.name == \"$INSTANCE\") | .spec.parameters" | yq -e --yaml-output '.' > zookeeper-params.yml
+```
+
+The above command generates a file called `zookeeper-params.yml` with the current values of all the parameters in effect for the `zookeeper` instance.
+
+### Updating Parameters
+Parameters can be updated using arguments to the KUDO CLI.
+
+**Example**: Increasing Zookeeper node counts
+- Increase the number of nodes using the KUDO CLI:
+```
+kubectl kudo update --instance zookeeper -p NODE_COUNT=5 -n test-project-zc6tc
+```
+- Monitor the KUDO Zookeeper deployment plan:
+```
+kubectl kudo plan status --instance zookeeper -n test-project-zc6tc
+```
+- Wait for the deployment plan to have a status of `COMPLETE`
+
+When the deployment plan is `COMPLETE` there should be 5 nodes as seen by the number of pods running:
+```
+$ kubectl get pods -n test-project-zc6tc
+NAME                    READY   STATUS    RESTARTS   AGE
+zookeeper-zookeeper-0   1/1     Running   0          37s
+zookeeper-zookeeper-1   1/1     Running   0          2m47s
+zookeeper-zookeeper-2   1/1     Running   0          3m48s
+zookeeper-zookeeper-3   1/1     Running   0          5m15s
+zookeeper-zookeeper-4   1/1     Running   0          5m15s
+```
+
+**Example**: Updating multiple parameters:
+
+To update multiple parameters at once, it is recommended to submit the updated parameters using the KUDO CLI.
+
+See [Available Parameters](#available-parameters) to get the full list of current parameters as a file.
+
+Apply the desired updates in `zookeeper-params.yml` using the KUDO CLI:
+```
+kubectl kudo update -n test-project-zc6tc --instance=zookeeper -P zookeeper-params.yml 
+```
+Wait for the deployment plan to `COMPLETE` as shown in the node counts example.
+
+### Limitations
+
+The following parameters are viewed as immutable by the operator after an instance has been deployed, any modifications to them will trigger the `not-allowed` plan:
+- `DISK_SIZE`
+- `STORAGE_CLASS`
+
+While modifying the KUDO Zookeeper parameters for `DISK_SIZE` are not permitted, users can resize the associated [PVC](https://kubernetes.io/blog/2018/07/12/resizing-persistent-volumes-using-kubernetes/).
+Switching between persistent & ephemeral storage classes is not supported for `STORAGE_TYPE`.
+
+### External Access
+
+KUDO Zookeeper creates two Kubernetes Services:
+- [Client Service (CS)](https://github.com/kudobuilder/operators/blob/master/repository/zookeeper/operator/templates/services.yaml#L20-L34) - For connecting clients to Zookeeper.
+- [High Availability Services (HS)](https://github.com/kudobuilder/operators/blob/master/repository/zookeeper/operator/templates/services.yaml#L1-L18) - For connecting additional Zookeeper servers to the quorum in [replicated mode](https://zookeeper.apache.org/doc/current/zookeeperStarted.html).
```

Comment:
I think it would be good to elaborate a bit here beyond just pointing at YAML, first explaining what the difference is between the 'CS' and the 'HS' services and then explaining precisely how the names translate to endpoints.   The Kudo Zookeeper docs do not seem to discuss this at all...

The example below mentions you use the 'CS' with the zkCLI but there is no mention of what the HS service is, and the YAML does not explain it either.  From experience, I believe you use the HS for the Zookeeper URI access string for connecting with Kafka/Cassandra, etc, and the URI looks something like this:

`<instance>-zookeeper-[0-n].<instance>-hs:2181` (??)

But we should document this here to make it clear and to avoid frustration as most people will probably rename these things from the default.

Reply:
@cneth well, there's two Services:
- One for the clients to talk to the ZK Quorum over port 2181.
- One for ZK servers to talk to each other over two other ports: 2888 for servers to listen to requests from each other, 3888 where leader election happens.

There's a mismatch between the docs and code on the KUDO Site:
`<instance>-zookeeper-[0-n].<instance>-hs:2181` doesn't line up with https://github.com/kudobuilder/operators/blob/master/repository/zookeeper/operator/templates/services.yaml#L20-L34

As it should be `<instance>-zookeeper-[0-n].<instance>-cs:2181`

Overall I understand where the confusion is coming from and I'll attempt to clarify this. Stay tuned.